rapidsai/legate-boost
# add conda builds

Contributes to #115

Adds conda builds.

## Notes for Reviewers

This is not even close to ready for review. Just putting it up here to show the direction I'm going. This diff will get smaller as other PRs are merged.

### Design choices

#### using GitHub Artifact store instead of RAPIDS S3 bucket

<details><summary>details (click me)</summary>

RAPIDS is looking to eventually move away from its S3-based strategy for storing and hosting CI artifacts. I wanted to use this as a test case for that.

GitHub Artifacts is also nice because, unlike https://downloads.rapids.ai/, it's on the public internet... so any outside-of-NVIDIA contributors can see their CI artifacts.

</details>

#### building `*_gpu` and `*_cpu*` variants

<details><summary>details (click me)</summary>

`legate-core` and `cunumeric` do something very similar. `legate-core` / `cunumeric` did not want to add a `*_gpu` portion to the build string when I proposed it, because they considered changing the build string format to be an unacceptable breaking change.

Since these packages for `legate-boost` are brand new, I think we *should* do that here, to achieve the following semantics:

```shell
# only installs the GPU variant (or fails if it's not installable)
conda install legate-core=24.06=*_gpu

# only installs the CPU variant (or fails if it's not installable)
conda install legate-core=24.06=*_cpu

# installs the GPU variant if CUDA is detected (look for '__cuda' in the output of 'conda info'),
# otherwise installs the CPU variant
conda install legate-core
```

</details>

#### Building and testing in separate environments

<details><summary>details (click me)</summary>

On `main`, this project's CI is building and testing in the same environment. This can lead to issues like those described in #144, where packaging problems are silently missed.

Building in one place and testing in another has a few nice benefits:

* less GPU runner utilization
  - *builds happen on CPU-only runner*
* CI is more likely to be able to detect packaging problems
  - *tests just install pre-built packages and all the dependencies they declare... could catch things like "this package needs `scipy` at runtime but it isn't in the package dependencies"*
* greater confidence in CPU support
  - *CPU tests run on a machine that doesn't have CUDA or a GPU*
  - *installation of the CPU-only variant tested there too*
* tests can run concurrently
  - *CPU tests and GPU tests run in physically separate instances, so can happen at the same time, which increases the amount of testing that can be done for any fixed amount of CI time*

</details>

### How to test this locally

See the notes added to `contributing.md` in the diff here (which also includes more explanation of these commands).

In short:

```shell
docker run \
  --rm \
  -v $(pwd):/opt/legate-boost \
  -w /opt/legate-boost \
  -it rapidsai/ci-conda:cuda12.5.1-ubuntu22.04-py3.11 \
  bash

gh auth login

RUN_ID=10584679029
gh run download \
    --dir "${RAPIDS_CONDA_BLD_OUTPUT_DIR}" \
    --repo rapidsai/legate-boost \
    --name "legate-boost-conda-cuda${RAPIDS_CUDA_VERSION}-amd64-py${PYTHON_VERSION}" \
    "${RUN_ID}"

rapids-dependency-file-generator \
  --output conda \
  --file-key test_python \
  --matrix "cuda=${RAPIDS_CUDA_VERSION%.*};arch=$(arch);py=${RAPIDS_PY_VERSION}" \
| tee /tmp/env.yaml

conda env create \
    --yes \
    --file /tmp/env.yaml \
    --name test-env

conda install \
  --name test-env \
  --channel "${RAPIDS_CONDA_BLD_OUTPUT_DIR}" \
  --channel legate \
  --channel conda-forge \
    'legate-boost=*=*_cpu'

source activate test-env

# run all the CPU tests
ci/run_pytests_cpu.sh

# or, just run one set of test cases
legate \
    --sysmem 28000 \
    --module pytest \
    legateboost/test/test_estimator.py::test_regressor \
    -sv \
    --durations=0
```
